server/src/main/java/org/apache/druid/segment/realtime/appenderator/UnifiedIndexerAppenderatorsManager.java

```diff
@@ -183,22 +192,48 @@ public Appenderator createOfflineAppenderatorForTask(
       IndexMerger indexMerger
   )
   {
-    Appenderator appenderator = Appenderators.createOffline(
-        schema,
-        rewriteAppenderatorConfigMemoryLimits(config),
-        metrics,
-        dataSegmentPusher,
-        objectMapper,
-        indexIO,
-        wrapIndexMerger(indexMerger)
-    );
-    return appenderator;
+    synchronized (this) {
```

Comment:
nit: it looks like the common code between `createRealtimeAppenderatorForTask` and `createOfflineAppenderatorForTask` can be extracted as a method which can handle all necessary synchronization and updating datasourceBundle.

Reply:
I decided to leave it as is, since the shared portion is pretty small here and I feel it's more readable without the additional indirection